### ISSUE TITLE
Migrate from github.com/docker/distribution to github.com/distribution/distribution/v3

### ISSUE THREAD

**Issue by howardjohn** [kubernetes/kubernetes#103849]
#### What happened:
Kubernetes libraries have a dependency on `github.com/docker/distribution`
#### What you expected to happen:
Kubernetes libraries have a dependency on `github.com/distribution/distribution/v3`. Note the github org just changed, its the same project. I believe the v2 -> v3 was mostly to get around go.mod weirdness with versions, rather than large breaking changes, but am not sure.
#### How to reproduce it (as minimally and precisely as possible):
Look at go.mod, I suppose

#### Anything else we need to know?:
`github.com/docker/distribution@v2.7.1+incompatible`, the library in use, is 2.5 years old. It does not have a go.mod file, and in turn depends on a bunch of other libraries without go.mod files. This generally makes importing these libraries a pain - you can see 350 projects needing to add `replace` hacks to go.mod to work around these issues: https://github.com/search?q=%22replace+github.com%2Fdocker%2Fdistribution%22&type=code

It would be great to upgrade to the latest version.

**Comment by dims** [comment]
@howardjohn `docker/docker` needs to go first :(

```
[dims@dims-a02 12:18] ~/go/src/k8s.io/kubernetes ⟩ rg '"github.com/docker/distribution'
pkg/kubelet/dockershim/libdocker/helpers.go
25:	dockerref "github.com/docker/distribution/reference"

pkg/kubelet/images/image_manager.go
23:	dockerref "github.com/docker/distribution/reference"

pkg/util/parsers/parsers.go
26:	dockerref "github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/image_pull.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/plugin_install.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/service_create.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/image_tag.go
7:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/plugin_upgrade.go
8:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/container_commit.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/image_create.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/image_push.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/client/image_import.go
9:	"github.com/docker/distribution/reference"

vendor/github.com/docker/docker/errdefs/http_helpers.go
8:	"github.com/docker/distribution/registry/api/errcode"

vendor/github.com/docker/distribution/reference/normalize.go
8:	"github.com/docker/distribution/digestset"

staging/src/k8s.io/kubectl/pkg/cmd/run/run.go
24:	"github.com/docker/distribution/reference"

staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
25:	"github.com/docker/distribution/reference"
```

**Comment by howardjohn** [comment]
I think https://github.com/kubernetes/kubernetes/pull/103850 is sufficient to drop the *direct* dependency at least. The indirect one... will require probably changes to probably 20+ repos

**Comment by dims** [comment]
@howardjohn i am happy to wait ... to avoid having 2 things that do the same thing in our dependencies.

**Comment by howardjohn** [comment]
Is dockershim removal still planned for 1.23 (which I think the `master` branch is now 1.23?) https://kubernetes.io/blog/2020/12/02/dockershim-faq/#when-will-dockershim-be-removed? Or do we need to wait a bit longer.

This is not a very high priority for us so I don't mind waiting months or years really

**Comment by pacoxu** [comment]
Pending on https://github.com/kubernetes/enhancements/issues/2221

**Comment by dims** [comment]
@pacoxu @howardjohn we are ready to move on this now, as dockershim removal is done!

**Comment by pacoxu** [comment]
https://github.com/distribution/distribution/issues/3587 is open, but it seems to be no update.

https://github.com/kubernetes/utils/pull/232 is closed.

**Comment by pacoxu** [comment]
According to @thaJeztah 's new PR  https://github.com/containerd/containerd/pull/9034, I migrate it to https://github.com/distribution/referenc v0.5.0 with https://github.com/kubernetes/kubernetes/pull/120332.

**Comment by pacoxu** [comment]
Finally we used `[github.com/distribution/referenc](https://github.com/distribution/referenc)` instead.

/close 
Per https://github.com/kubernetes/kubernetes/pull/120316. Thanks @dims @thaJeztah @Jamstah.
